row_transform/row_transform.tex

```diff
@@ -0,0 +1,188 @@
+        
+        \item \textup{(Stability guarantee).} For every pair of elements $v, w$ in \texttt{input\_domain} and for every pair $(\din, \dout)$, where $\din$ is of the associated type for \texttt{input\_metric} and $\dout$ is the associated type for \texttt{output\_metric}, if $v,w$ are $\din$-close under \texttt{input\_metric} and $\Relation(\din, \dout) = \True$, then $\function(v), \function(w)$ are $\dout$-close under \texttt{output\_metric}.
+    \end{enumerate}
+\end{theorem}
+\begin{proof}
+\begin{enumerate}
+\item \textbf{(Appropriate output domain).} In the case of \texttt{make\_row\_by\_row}, this corresponds to showing that for every vector $v$ of elements of type \texttt{DIA}, \texttt{function}$(v)$ is a vector of elements of type \texttt{DOA}. 
+
+The $\function(v)$ has type \texttt{Vec[DOA]} follows from the assumption that element $v$ is in \texttt{input\_domain} and from the type signature of \texttt{function} in line~\ref{line:fn} of the pseudocode (Section~\ref{sec:pseudocode}), which takes in an element of type \texttt{Vec(DIA)} and returns an element of type \texttt{Vec[DOA]}. If the Rust code compiles correctly, then the type correctness follows from the definition of the type signature enforced by Rust. Otherwise, the code raises a compile time error for incorrect function input type or output type. 
+
+Notice also that the appropriate output domain is achieved during run time since the \texttt{atom\_function} is a pure function with a function trait that maps elements of type \texttt{\&DIA::Carrier} to \texttt{DOA::Carrier}. Therefore the  \texttt{function} correctly maps elements in the input domain \texttt{VectorDomain(DIA)} to the output domain of \texttt{VectorDomain(DOA)} during run time.
+
+
+\item \textbf{(Domain-metric compatibility).} The Symmetric distance is both the \texttt{input\_metric} and \texttt{output\_metric}. Symmetric distance is compatible with \texttt{VectorDomain(T)} for any generic type \texttt{T}, as stated in \href{https://www.overleaf.com/project/60d215bf90b337ac02200a99}{``List of definitions used in the pseudocode"}. The theorem holds because for \texttt{make\_row\_by\_row}, the input domain is \texttt{VectorDomain(DIA)} and the output domain is \texttt{VectorDomain(DOA)}. 
+
+
+\item \textbf{(Stability guarantee).} 
+Recall that \texttt{function} is a row transformation with respect to pure function \texttt{atom\_function}, which we denote as $f$ for simplicity. We want to show that
+
+$$d_{Sym}(\texttt{function}(v), \texttt{function}(w)) \leq d_{Sym}(v, w).$$
+
+We use the histogram notation. Recall that $h_{\texttt{function}(v)}(z)$ is the number of occurrences of $z$ in vector \texttt{function}$(v)$. This is equivalent to the sum of the number of occurrences of each $y \in f^{(-1)}(z)$ in vector $v$ since $f$ is a pure function. \grace{Does $f$ have to be onto function?} Since $h_{\texttt{function}(v)}(z) = \sum_{y \in f^{-1}(z)} h_v(y)$, we have:
```

Comment:
For stability guarantee: Does f has to be onto function, or pure function assumption is sufficient in order to make this claim that $h_{function(v)}(z) = \sum_{y \in f^{-1}(z)} h_v(y)$?

Reply:
Consider `Function(|x: i8| true)` that maps i8 -> bool. This is row-by-row, but not onto (because no input x can map to false).

Reply:
No, f does not have to be onto.  If z is outside the image of f, then both sides are equal to 0.

Reply:
But in the sentence `This is equivalent to the sum of the number of occurrences of each $y \in f^{(-1)}(z)$ in vector $v$ since $f$ is a pure function`, why do we need the assumption that `f` is a pure function? It would be good to separate the proof strategy that simplifies many of the proofs (e.g., clamp) vs the code actually calling the row-by-row transformation. I think that in the former, we do not need the pure function requirement, since the code is not calling a different transformation, but in the latter we do. However, as it is right now, they seem to be intertwined, and then it does not follow from the Lemma that I can simplify clamp only in the proof without actually having to change the code.